### ISSUE TITLE
ImageData DataArrays Flipped

### ISSUE THREAD

**Issue by jaredmagnusson** [pyvista/pyvista#4862]
### Describe the bug, what's wrong, and what you expected.

Plotting or saving ImageData changes how the data is stored. For example, if I plot a YZ slice of my dataset, pyvista transposes the data array such that it no longer matches the original ordering. I saw the same behavior using a RectilinearGrid.

### Steps to reproduce the bug.

```python
import pyvista
import numpy as np
import matplotlib.pyplot as plt

def main():
    # create grid
    dimension = (1,5,3)
    origin=(0,0,0)
    spacing=(0,1,1)
    grid = pyvista.ImageData(
        dimensions=dimension,
        spacing=spacing,
        origin=origin,
    )
    # initialize dummy data - vary in Y, constant in Z
    arr = np.expand_dims(np.array([1,2,3,4,5]),axis=1)
    V = np.zeros(dimension)
    V[:] = np.tile(arr, dimension[2])
    grid["V"] = V.ravel()
    # create meshgrid for matplotlib
    y = np.linspace(0,spacing[1]*(dimension[1]-1), dimension[1])
    z = np.linspace(0,spacing[2]*(dimension[2]-1), dimension[2])
    Y,Z = np.meshgrid(y,z,indexing="ij")
    # plot via matplotlib
    plt.figure()
    plt.contourf(Y,Z,grid["V"].reshape(dimension[1],dimension[2]))
    plt.axis("equal")
    # plt.tight_layout()
    plt.xlim([0,4])
    plt.xlabel("y")
    plt.ylim([0,2])
    plt.ylabel("z")
    plt.title("Matplotlib")
    plt.savefig("grid_imagedata_mplotlib.png")
    # save image data mesh
    grid.save("grid_imagedata.vti")

if __name__=="__main__":
    main()
```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Thu Sep 07 10:37:19 2023 CDT

                OS : Linux
            CPU(s) : 12
           Machine : x86_64
      Architecture : 64bit
       Environment : Python
       GPU Details : error
  MathText Support : False

  Python 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC
  12.2.0]

           pyvista : 0.41.1
               vtk : 9.2.6
             numpy : 1.25.2
        matplotlib : 3.7.2
            scooby : 0.7.2
             pooch : v1.7.0
             scipy : 1.11.2
            meshio : 5.3.4
--------------------------------------------------------------------------------
```

### Screenshots

Expected: plotting variable "V" as a contourf plot in matplotlib. The data is varying in Y and is constant in Z.
![matplotlib](https://github.com/pyvista/pyvista/assets/87670084/7940c6cd-8526-4207-b2f1-9432906725c3)
Actual: plotting variable "V" in Paraview (saw same behavior with pyvista.Plotter). The Y and Z position data is correct, but dataset "V" has been transposed or otherwise jumbled.
![pyvista_render_paraview](https://github.com/pyvista/pyvista/assets/87670084/b9342113-c6e6-4b73-8de7-e71331a2f64f)

**Comment by jaredmagnusson** [comment]
I found my answer - ImageData data arrays are stored in Fortran order. See [Creating A Uniform Grid](https://tutorial.pyvista.org/tutorial/02_mesh/solutions/c_create-uniform-grid.html). By setting my data arrays as
```python
grid["V"] = V.flatten(order="F")
```
both matplotlib and paraview/pyvista agreed. To access the data again as a numpy array, use
```python
grid["V"].reshape(*dimensions, order="F")
```

**Comment by UnaiSan** [comment]
This happens in VTK, and it is sometimes inconvenient when going to and from Numpy N-dimensional arrays (`ravel`, `flatten`, etc take order="C" as default).

It would be nice if we could pass directly the N-dimensional array and pyvista would flatten and unflatten for us

for example, if we have a (N1, N2, N3) structured mesh, so the coordinates of the points have (N1, N2, N3) shape, and we do

```{python}
u1 = np.linspace(0, 1, N1)
u2 = np.linspace(0, 1, N2)
u3 = np.linspace(0, 1, N3)
U1, U2, U3 = np.meshgrid(u1, u2, u3, indexing="ij")
X = f_x(u1, u2, u3)   # shape (N1, N2, N3)
Y = f_y(u1, u2, u3)   # shape (N1, N2, N3)
Z = f_z(u1, u2, u3)   # shape (N1, N2, N3)
C = f_c(X, Y, Z)         # shape (N1, N2, N3)

mesh = pyvista.StructuredGrid(X, Y, Z)
```

Then it would be nice to be able to do
```
mesh.point_data["some_data"] = C
```
instead of

```
mesh.point_data["some_data"] = C.ravel(order="F")
```

**Comment by jaredmagnusson** [comment]
Exactly, I ended up writing a wrapper class around ImageData to handle those getter/setter inconveniences. It'd also be nice if ImageData had a way to get x,y,z 1D arrays, as well as meshgrid attributes like RectilinearGrid currently enjoys. It's a more convenient interface than working with `.points`

**Comment by MatthewFlamm** [comment]
> Then it would be nice to be able to do
> 
> ```
> mesh.point_data["some_data"] = C
> ```
> 
> instead of
> 
> ```
> mesh.point_data["some_data"] = C.ravel(order="F")
> ```

There are several cases that need to be considered.  Here are a few":

- What if the user is storing their data in `C` order format?
- What about vector valued data?
  - Vector valued data cannot be raveled directly.
  - Edge cases also arise when dimensionality of the mesh overlaps with the dimensions of the vector
  - Is the data being passed in a vector or part of a structured mesh dataset?

What if the mesh isn't Structured?  If we special case for structured data, this makes for a messy API.  I'm not trying to say this cannot be done, but we'd have to think through all of these cases before implementing it.

**Comment by user27182** [comment]
An API for automatically reshaping arrays is being proposed in #7100. @jaredmagnusson @UnaiSan I would like your feedback on this API and would like to accommodate your reshaping use-case and possibly include it as part of this new API. Please see [this comment](https://github.com/pyvista/pyvista/issues/7098#issuecomment-2602871809) for more details about the feedback that I am looking for.

I _think_ the following example may already be covered by the `dimensions` reshaping method and `ravel_f` flattening methods already included so far with PR #7100 , but please have a look and let me know.

> It would be nice if we could pass directly the N-dimensional array and pyvista would flatten and unflatten for us
> 
> for example, if we have a (N1, N2, N3) structured mesh, so the coordinates of the points have (N1, N2, N3) shape, and we do
> 
> ```
> u1 = np.linspace(0, 1, N1)
> u2 = np.linspace(0, 1, N2)
> u3 = np.linspace(0, 1, N3)
> U1, U2, U3 = np.meshgrid(u1, u2, u3, indexing="ij")
> X = f_x(u1, u2, u3)   # shape (N1, N2, N3)
> Y = f_y(u1, u2, u3)   # shape (N1, N2, N3)
> Z = f_z(u1, u2, u3)   # shape (N1, N2, N3)
> C = f_c(X, Y, Z)         # shape (N1, N2, N3)
> 
> mesh = pyvista.StructuredGrid(X, Y, Z)
> ```
> 
> Then it would be nice to be able to do
> 
> ```
> mesh.point_data["some_data"] = C
> ```
> 
> instead of
> 
> ```
> mesh.point_data["some_data"] = C.ravel(order="F")
> ```

**Comment by jaredmagnusson** [comment]
Thank you for including me on this. If there was an ImageData dataset with one of its dimensions equal to 1, would this cause a problem with cell data reshaping? I'm looking at `_reshape_array` around line 1572:

```python
def _reshape_array(self: Self, array: pyvista_ndarray) -> pyvista_ndarray:
        if self._reshape_method == 'dimensions':
            if hasattr(self.dataset, 'dimensions'):
                dims = self.dataset.dimensions
                newshape = dims
                if self.association == FieldAssociation.CELL:
                    newshape -= 1
```

I'm just wondering if cell data having a dimension of 0 is a problem

**Comment by user27182** [comment]
> Thank you for including me on this. If there was an ImageData dataset with one of its dimensions equal to 1, would this cause a problem with cell data reshaping? I'm looking at `_reshape_array` around line 1572:
> ``` python
> def _reshape_array(self: Self, array: pyvista_ndarray) -> pyvista_ndarray:
>         if self._reshape_method == 'dimensions':
>             if hasattr(self.dataset, 'dimensions'):
>                 dims = self.dataset.dimensions
>                 newshape = dims
>                 if self.association == FieldAssociation.CELL:
>                     newshape -= 1
>```
> I'm just wondering if cell data having a dimension of 0 is a problem

Thanks for the feedback, yes these are good edge cases which will need to be accommodated and tested. If the data is 2D (one dimension is `1`), then maybe we also want to squeeze the array so that we have an array shape `(N, M)` instead of `(N, M, 1)`. Though it's possible that this dimension may want to be kept as well. 

I am wondering if an intermediate `ArrayReshaper` class or something similar would be better to have to specify all of these options instead.

E.g. do `ArrayReshaper(flattening_method=..., reshaping_method=..., squeeze: bool=False, order: str='C', other_options=...)`

**Comment by jaredmagnusson** [comment]
yeah definitely providing the option would be needed - sometimes the user might want to treat such data consistently in their operations (always with dimension=3), and other times if they're constantly updating their datasets from kernels that operate on 2D arrays, the squeeze could be convenient
